### PR TITLE
Convert all packages to BigInt instead of bn.js

### PR DESCRIPTION
**Pull request by ec2**

### CODE REVIEW

**Comment by wemeetagain** [comment]
:heart_eyes: :100: :boom:

File: package.json [hunk 1]
```diff
@@ -45,5 +45,8 @@
   },
   "optionalDependencies": {
     "@chainsafe/eth2-spec-tests": "0.8.3"
+  },
+  "dependencies": {
+    "@babel/plugin-syntax-bigint": "^7.4.4"
```
Comment (wemeetagain) [inline package.json]: dev dependency

File: packages/eth2.0-params/src/presets/minimal.ts [hunk 2]
```diff
@@ -20,10 +18,10 @@ export const DEPOSIT_CONTRACT_ADDRESS = 0;
 export const DEPOSIT_CONTRACT_TREE_DEPTH = 2 ** 5; // 32
 
 // Gwei Values
-export const MIN_DEPOSIT_AMOUNT = new BN("1000000000"); // 1,000,000,000 Gwei
-export const MAX_EFFECTIVE_BALANCE = new BN("32000000000"); // 32,000,000,000 Gwei
-export const EJECTION_BALANCE = new BN("16000000000"); // 16,000,000,000 Gwei
-export const EFFECTIVE_BALANCE_INCREMENT = new BN("1000000000"); // 1,000,000,000 Gwei
+export const MIN_DEPOSIT_AMOUNT = BigInt("1000000000"); // 1,000,000,000 Gwei
```
Comment (wemeetagain) [inline packages/eth2.0-params/src/presets/minimal.ts:21]: I realize that these were always in a different format than the mainnet params, but consider adjusting to be more similar. Maybe we save that for another PR

File: packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts [hunk 3]
```diff
@@ -49,11 +48,9 @@ export function getAttestationDeltas(config: IBeaconConfig, state: BeaconState):
       const attestingBalance = getTotalBalance(state, unslashedAttestingIndices);
       eligibleValidatorIndices.forEach((index) => {
         if (unslashedAttestingIndices.includes(index)) {
-          rewards[index] = rewards[index]
-            .add(getBaseReward(config, state, index).mul(attestingBalance).div(totalBalance));
+          rewards[index] = rewards[index] + (getBaseReward(config, state, index) * attestingBalance / totalBalance);
         } else {
-          penalties[index] = penalties[index]
-            .add(getBaseReward(config, state, index));
+          penalties[index] = penalties[index] + getBaseReward(config, state, index);
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts]: you can use += here and on L51

File: packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts [hunk 4]
```diff
@@ -63,17 +60,16 @@ export function getAttestationDeltas(config: IBeaconConfig, state: BeaconState):
       .filter((a) => getAttestingIndices(config, state, a.data, a.aggregationBits).includes(index))
       .reduce((a1, a2) => a2.inclusionDelay < a1.inclusionDelay ? a2 : a1);
     const baseReward = getBaseReward(config, state, index);
-    const proposerReward = baseReward.divn(config.params.PROPOSER_REWARD_QUOTIENT);
-    rewards[earliestAttestation.proposerIndex] = rewards[earliestAttestation.proposerIndex]
-      .add(proposerReward);
-    const maxAttesterReward = baseReward.sub(proposerReward);
+    const proposerReward = baseReward / BigInt(config.params.PROPOSER_REWARD_QUOTIENT);
+    rewards[earliestAttestation.proposerIndex] = rewards[earliestAttestation.proposerIndex] + proposerReward;
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts]: +=

File: packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts [hunk 5]
```diff
@@ -63,17 +60,16 @@ export function getAttestationDeltas(config: IBeaconConfig, state: BeaconState):
...
+    const maxAttesterReward = baseReward - proposerReward;
     rewards[index] = rewards[index]
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts]: +=

File: packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts [hunk 6]
```diff
@@ -82,14 +78,12 @@ export function getAttestationDeltas(config: IBeaconConfig, state: BeaconState):
     const matchingTargetAttestingIndices =
       getUnslashedAttestingIndices(config, state, matchingTargetAttestations);
     eligibleValidatorIndices.forEach((index) => {
-      penalties[index] = penalties[index]
-        .add(getBaseReward(config, state, index).muln(config.params.BASE_REWARDS_PER_EPOCH));
+      penalties[index] = (penalties[index] + getBaseReward(config, state, index))
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts]: +=

File: packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts [hunk 7]
```diff
@@ -82,14 +78,12 @@ export function getAttestationDeltas(config: IBeaconConfig, state: BeaconState):
...
+        * BigInt(config.params.BASE_REWARDS_PER_EPOCH);
       if (!matchingTargetAttestingIndices.includes(index)) {
-        penalties[index] = penalties[index]
-          .add(
-            state.validators[index].effectiveBalance
-              .muln(finalityDelay)
-              .div(config.params.INACTIVITY_PENALTY_QUOTIENT));
+        penalties[index] = penalties[index] + (
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/src/epoch/balanceUpdates/attestation.ts]: +=

File: packages/eth2.0-state-transition/src/epoch/balanceUpdates/crosslink.ts [hunk 8]
```diff
@@ -35,10 +34,9 @@ export function getCrosslinkDeltas(config: IBeaconConfig, state: BeaconState): [
     crosslinkCommittee.forEach((index) => {
       const baseReward = getBaseReward(config, state, index);
       if (attestingIndices.includes(index)) {
-        rewards[index] = rewards[index]
-          .add(baseReward.mul(attestingBalance).div(committeeBalance));
+        rewards[index] = rewards[index] + (baseReward * attestingBalance / committeeBalance);
       } else {
-        penalties[index] = penalties[index].add(baseReward);
+        penalties[index] = penalties[index] + baseReward;
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/src/epoch/balanceUpdates/crosslink.ts]: += here and above

File: packages/eth2.0-state-transition/src/epoch/balanceUpdates/baseReward.ts [hunk 9]
```diff
@@ -15,6 +15,8 @@ export function getBaseReward(
 ): Gwei {
   const totalBalance = getTotalActiveBalance(config, state);
   const effectiveBalance = state.validators[index].effectiveBalance;
-  return effectiveBalance.muln(config.params.BASE_REWARD_FACTOR)
-    .div(bnSqrt(totalBalance)).divn(config.params.BASE_REWARDS_PER_EPOCH);
+  return effectiveBalance
+    * BigInt(config.params.BASE_REWARD_FACTOR)
+    / bnSqrt(totalBalance)
+    / BigInt(config.params.BASE_REWARDS_PER_EPOCH);
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/src/epoch/balanceUpdates/baseReward.ts:21]: I feel like BASE_REWARD_FACTOR and BASE_REWARDS_PER_EPOCH could be bigints. Maybe for another PR

File: packages/eth2.0-state-transition/src/util/validatorStatus.ts [hunk 10]
```diff
@@ -65,19 +65,19 @@ export function slashValidator(
 
   const slashedBalance = state.validators[slashedIndex].effectiveBalance;
   state.slashings[currentEpoch % config.params.EPOCHS_PER_SLASHINGS_VECTOR] =
-    state.slashings[currentEpoch % config.params.EPOCHS_PER_SLASHINGS_VECTOR].add(slashedBalance);
+    state.slashings[currentEpoch % config.params.EPOCHS_PER_SLASHINGS_VECTOR] + slashedBalance;
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/src/util/validatorStatus.ts]: +=

File: packages/eth2.0-state-transition/src/util/validatorStatus.ts [hunk 11]
```diff
@@ -65,19 +65,19 @@ export function slashValidator(
...
 
 
   const proposerIndex = getBeaconProposerIndex(config, state);
   if (whistleblowerIndex === undefined || whistleblowerIndex === null) {
     whistleblowerIndex = proposerIndex;
   }
-  const whistleblowingReward = slashedBalance.divn(config.params.WHISTLEBLOWING_REWARD_QUOTIENT);
-  const proposerReward = whistleblowingReward.divn(config.params.PROPOSER_REWARD_QUOTIENT);
+  const whistleblowingReward = slashedBalance / BigInt(config.params.WHISTLEBLOWING_REWARD_QUOTIENT);
+  const proposerReward = whistleblowingReward/ BigInt(config.params.PROPOSER_REWARD_QUOTIENT);
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/src/util/validatorStatus.ts:79]: these REWARD_QUOTIENTs could potentially be bigints too

File: packages/eth2.0-state-transition/test/unit/stateTransition/block/operations/deposit.test.ts [hunk 12]
```diff
@@ -85,7 +84,7 @@ describe("process block - deposits", function () {
     const state = generateState({eth1DepositIndex: 3});
     verifyMerkleBranchStub.returns(true);
     const deposit = generateDeposit();
-    deposit.data.amount = new BN(config.params.MAX_EFFECTIVE_BALANCE);
+    deposit.data.amount = BigInt(config.params.MAX_EFFECTIVE_BALANCE);
```
Comment (wemeetagain) [inline packages/eth2.0-state-transition/test/unit/stateTransition/block/operations/deposit.test.ts]: I think MAX_EFFECTIVE_BALANCE is already a bigint

File: packages/eth2.0-utils/src/bytes.ts [hunk 13]
```diff
@@ -1,24 +1,21 @@
...
-    return value.toArrayLike(Buffer, "le", length);
+  } else { // value is number and is too large, or a BigInt
+    value = BigInt(value);
+    return toBufferLE(value, length);
   }
 }
 
-export function bytesToBN(value: bytes): BN {
-  return new BN(value, "le");
+export function bytesToBN(value: bytes): bigint {
```
Comment (wemeetagain) [inline packages/eth2.0-utils/src/bytes.ts]: can you find/replace with `bytesToBigint` or `bytesToBigInt`

File: packages/eth2.0-utils/src/math.ts [hunk 14]
```diff
@@ -35,12 +34,12 @@ export function intSqrt(n: number): number {
...
-  let x = n.clone();
-  let y = x.addn(1).divn(2);
-  while (y.lt(x)) {
+export function bnSqrt(n: bigint): bigint {
+  let x = n;
+  let y = (x + 1n)/ 2n;
+  while (y < x) {
     x = y;
-    y = x.add(n.div(x)).divn(2);
+    y = (x + n/x)/ 2n;
```
Comment (wemeetagain) [inline packages/eth2.0-utils/src/math.ts]: add spacing

File: packages/eth2.0-utils/src/math.ts [hunk 15]
```diff
@@ -2,20 +2,19 @@
...
  */
 
-import BN from "bn.js";
 
 /**
  * Return the min number between two big numbers.
  */
-export function bnMin(a: BN, b: BN): BN {
-  return a.lt(b) ? a : b;
+export function bnMin(a: bigint, b: bigint): bigint {
```
Comment (wemeetagain) [inline packages/eth2.0-utils/src/math.ts]: can we call these functions `bigint*` or `bigInt*`

File: packages/eth2.0-utils/src/yaml/int.ts [hunk 16]
```diff
@@ -119,43 +118,48 @@ function constructYamlInteger(data: string): BN {
...
 
-  if (value === "0") return new BN(0);
+  if (value === "0") return 0n;
 
   if (ch === "0") {
-    if (value[1] === "b") return (new BN(value.slice(2), 2)).muln(sign);
-    if (value[1] === "x") return (new BN(value, 16)).muln(sign);
-    return (new BN(value, 8)).muln(sign);
+    if (value[1] === "b") return BigInt("0b" + value.slice(2)) * BigInt(sign);
+    if (value[1] === "x") return BigInt( value) * BigInt(sign);
```
Comment (wemeetagain) [inline packages/eth2.0-utils/src/yaml/int.ts]: spacing,
I think these two if statements can be collapsed to
`if (value[1] === "b" || value[1] === "x") return BigInt(value) * BigInt(sign);`

File: packages/eth2.0-utils/src/yaml/int.ts [hunk 17]
```diff
@@ -119,43 +118,48 @@ function constructYamlInteger(data: string): BN {
...
-  return (new BN(value, 10)).muln(sign);
+  return BigInt(value) * BigInt(sign);
 }
 
 function isInteger(object: object): boolean {
-  return BN.isBN(object);
+  try{
+    BigInt(object);
+    return true;
+  }  catch (e){
```
Comment (wemeetagain) [inline packages/eth2.0-utils/src/yaml/int.ts]: spacing

File: packages/eth2.0-utils/src/yaml/int.ts [hunk 18]
```diff
@@ -119,43 +118,48 @@ function constructYamlInteger(data: string): BN {
...
   }
 
-  return (new BN(value, 10)).muln(sign);
+  return BigInt(value) * BigInt(sign);
 }
 
 function isInteger(object: object): boolean {
-  return BN.isBN(object);
+  try{
+    BigInt(object);
```
Comment (wemeetagain) [inline packages/eth2.0-utils/src/yaml/int.ts]: I think to get the same behavior we need to do a typeof check here for "bigint"

File: packages/lodestar/src/util/bytes.ts [hunk 19]
```diff
@@ -2,29 +2,26 @@
...
-    value = new BN(value);
-    return value.toArrayLike(Buffer, "le", length);
+    value = BigInt(value);
+    return toBufferLE(value, length);
   }
 }
 
-export function bytesToBN(value: bytes): BN {
-  return new BN(value, "le");
+export function bytesToBN(value: bytes): bigint{
```
Comment (wemeetagain) [inline packages/lodestar/src/util/bytes.ts]: can you rename to `bytesToBigint` or `bytesToBigInt`?

File: packages/lodestar/test/utils/misc.ts [hunk 20]
```diff
@@ -9,8 +8,8 @@ export function randBetween(min: number, max: number): number {
 
 /**
  * Wraps randBetween and returns a bigNumber.
- * @returns {BN}
+ * @returns {bigint}
  */
-export function randBetweenBN(min: number, max: number): BN {
-  return new BN(randBetween(min, max));
+export function randBetweenBN(min: number, max: number): bigint {
```
Comment (wemeetagain) [inline packages/lodestar/test/utils/misc.ts]: `randBetweenBigint` or `randBetweenBigInt`

File: packages/ssz-cli/.vscode/settings.json [hunk 21]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (wemeetagain) [inline packages/ssz-cli/.vscode/settings.json]: I think this whole directory gotta go

File: packages/ssz/src/core/assertValidValue.ts [hunk 22]
```diff
@@ -47,12 +46,12 @@ export function assertValidValue(value: any, type: AnySSZType): void {
...
     case Type.uint:
-      assert(BN.isBN(value) || value === Number(value), "Invalid uint: not a uint");
+      assert( typeof value === "bigint" || typeof value === "number", "Invalid uint: not a uint");
       if (value === Infinity) {
         break;
       }
-      assert((new BN(value)).gten(0), "Invalid uint: value < 0");
-      assert((new BN(value)).lt((new BN(2)).pow(new BN(type.byteLength * 8))), "Invalid uint: not in range");
+      assert(value >= 0, "Invalid uint: value < 0");
+      assert(value < (BigInt(2))**(BigInt(type.byteLength * 8)), "Invalid uint: not in range");
```
Comment (wemeetagain) [inline packages/ssz/src/core/assertValidValue.ts]: I don't think there needs to be this many paren

File: packages/ssz/src/core/deserialize.ts [hunk 23]
```diff
@@ -152,8 +152,8 @@ function _deserializeUint(data: Buffer, type: UintType, start: number): Uint {
   if (type.byteLength > 6 && type.useNumber && uintData.equals(Buffer.alloc(type.byteLength, 255))) {
     return Infinity;
   } else {
-    const bn = new BN(uintData, 16, "le");
-    return (type.useNumber || type.byteLength <= 6) ? bn.toNumber() : bn;
+    const bn = BigInt("0x" + uintData.reverse().toString("hex"));
```
Comment (wemeetagain) [inline packages/ssz/src/core/deserialize.ts]: can you use `toBigIntLE`?
also change bn variable name

File: packages/ssz/test/unit/utils.ts [hunk 24]
```diff
@@ -1,5 +1,6 @@
 import {SimpleSSZType,} from "@chainsafe/ssz-type-schema";
-
+// @ts-ignore
+BigInt.prototype.toJSON = function() { return this.toString(); };
```
Comment (wemeetagain) [inline packages/ssz/test/unit/utils.ts]: why do we need this here? this doesn't seem right

File: packages/ssz/test/unit/deserialize.test.ts [hunk 25]
```diff
@@ -77,8 +77,8 @@ describe("deserialize", () => {
   for (const {value, type, expected} of testCases) {
     it(`should correctly deserialize ${stringifyType(type)}`, () => {
       const actual = deserialize(Buffer.from(value, "hex"), type);
-      if (BN.isBN(expected)) {
-        assert(expected.eq(actual as BN), `actual: ${actual}, expected: ${expected}`);
+      if (typeof expected == "bigint") {
+        assert(expected ==  actual, `actual: ${actual}, expected: ${expected}`);
```
Comment (wemeetagain) [inline packages/ssz/test/unit/deserialize.test.ts]: I think this condition can be removed, using the deepEqual in all cases

**Review by wemeetagain — CHANGES_REQUESTED** [review]
lotsa minor things

File: .eslintrc.js [hunk 26]
```diff
@@ -62,12 +65,12 @@ module.exports = {
     "new-parens": "error",
     "no-caller": "error",
     "no-bitwise": "off",
-    "no-conditional-assignment": true,
```
Comment (mpetrunic) [inline .eslintrc.js:65]: why did you remove this rule :(

File: packages/eth2.0-params/src/presets/minimal.ts (hunk 2 above)
Comment (mpetrunic) [reply packages/eth2.0-params/src/presets/minimal.ts:21]: https://github.com/ChainSafe/lodestar/issues/351 :disappointed:

File: packages/lodestar/src/db/api/validator/validator.ts [hunk 27]
```diff
@@ -59,7 +59,7 @@ export class ValidatorDB extends DatabaseService implements IValidatorDB {
   }
 
   private incrementPubKey(pubKey: BLSPubkey): string {
-    return new BN(pubKey).addn(1).toString("hex").replace("0x", "");
+    return (toBigIntLE(pubKey) + 1n).toString(16).replace("0x", "");
```
Comment (mpetrunic) [inline packages/lodestar/src/db/api/validator/validator.ts]: I think this should be bytesToBigint from utils so we don't need bigint-buffer dependency

File: packages/lodestar/src/db/schema.ts [hunk 28]
```diff
@@ -40,14 +38,14 @@ export enum Key {
...
+export function encodeKey(bucket: Bucket, key: Buffer | string | number | bigint, useBuffer = true): Buffer | string {
   let buf;
   if (typeof key === "string") {
     buf = Buffer.alloc(key.length + 1);
     buf.write(key, 1);
-  } else if (typeof key === "number" || BN.isBN(key)) {
+  } else if (typeof key === "number" || typeof key === 'bigint') {
     buf = Buffer.alloc(9);
-    (new BN(key)).toArrayLike(Buffer, "le", 8).copy(buf, 1);
+    toBufferLE(BigInt(key), 8).copy(buf, 1);
```
Comment (mpetrunic) [inline packages/lodestar/src/db/schema.ts]: intToBytes?

File: packages/lodestar/src/interop/keypairs.ts [hunk 29]
```diff
@@ -19,9 +19,8 @@ export function interopKeypairs(validatorCount: number): IKeypair[] {
 }
 
 export function interopKeypair(index: number): IKeypair {
-  const privkey = bytesToBN(hash(intToBytes(index, 32)))
-    .mod(CURVE_ORDER)
-    .toArrayLike(Buffer, "be", 32);
+  const privkey =
+    toBufferBE(bytesToBN(hash(intToBytes(index, 32))) % CURVE_ORDER, 32);
```
Comment (mpetrunic) [inline packages/lodestar/src/interop/keypairs.ts]: intToBytes?

File: packages/lodestar/src/util/bytes.ts [hunk 30]
```diff
@@ -2,29 +2,26 @@
  * @module util/bytes
```
Comment (mpetrunic) [inline packages/lodestar/src/util/bytes.ts]: this file is duplicate form eth2.0-utils and probably should be removed

File: packages/lodestar/test/e2e/sync/reqResp.test.ts [hunk 31]
```diff
@@ -47,7 +46,7 @@ describe("[sync] rpc", function () {
     const chain = new MockBeaconChain({
       genesisTime: 0,
       chainId: 0,
-      networkId: new BN(0),
+      networkId: 0n,
```
Comment (mpetrunic) [inline packages/lodestar/test/e2e/sync/reqResp.test.ts:49]: funny, why is network id bigint haha

File: packages/lodestar/test/unit/chain/forkChoice/attestationAggregator.test.ts [hunk 32]
```diff
@@ -8,17 +7,17 @@ import {
 
 
 describe("AttestationAggregator", () => {
-  const blockSlots = {
+  const blockSlots: any = {
```
Comment (mpetrunic) [inline packages/lodestar/test/unit/chain/forkChoice/attestationAggregator.test.ts]: {[k: string]: number}

File: packages/lodestar/test/unit/chain/forkChoice/attestationAggregator.test.ts [hunk 33]
```diff
@@ -8,17 +7,17 @@ import {
...
     "a": 1,
     "b": 1,
     "c": 1,
     "d": 2,
   }
-  const blockToSlot = (b) => blockSlots[b];
+  const blockToSlot: any = (b: any) => blockSlots[b];
```
Comment (mpetrunic) [inline packages/lodestar/test/unit/chain/forkChoice/attestationAggregator.test.ts]: b: string

at least make blockToSlot function type :)

File: packages/lodestar/test/unit/db/schema.test.ts [hunk 34]
```diff
@@ -19,10 +19,8 @@ describe("encodeKey", () => {
         expected = Buffer.concat([Buffer.from([bucket]), key as Buffer]);
       } else if (typeof key === "string") {
         expected = Buffer.concat([Buffer.from([bucket]), Buffer.from(key)]);
-      } else if (typeof key === "number") {
-        expected = Buffer.concat([Buffer.from([bucket]), (new BN(key)).toArrayLike(Buffer, 'le', 8)]);
-      } else if (BN.isBN(key)) {
-        expected = Buffer.concat([Buffer.from([bucket]), key.toArrayLike(Buffer, 'le', 8)]);
+      } else if (typeof key === "number" || typeof key === "bigint") {
+        expected = Buffer.concat([Buffer.from([bucket]), toBufferLE(BigInt(key), 8)]);
```
Comment (mpetrunic) [inline packages/lodestar/test/unit/db/schema.test.ts]: intToBytes

File: packages/ssz/src/core/deserialize.ts (hunk 23 above)
Comment (mpetrunic) [reply packages/ssz/src/core/deserialize.ts]: Not 100% but I think we ain't suppose to have eth2.0-utils dependency in ssz

File: packages/ssz/src/core/serialize.ts [hunk 35]
```diff
@@ -107,13 +107,13 @@ export function serialize(value: any, type: AnySSZType): Buffer {
...
+  let bnValue: bigint;
   if (type.byteLength > 6 && type.useNumber && value === Infinity) {
-    bnValue = new BN(Buffer.alloc(type.byteLength, 255));
+    bnValue = BigInt("0x" + Buffer.alloc(type.byteLength, 255).toString("hex"));
   } else {
-    bnValue = new BN(value);
+    bnValue = BigInt(value);
   }
-  bnValue.toArrayLike(Buffer, "le", type.byteLength)
+  toBufferLE(bnValue, type.byteLength)
```
Comment (mpetrunic) [inline packages/ssz/src/core/serialize.ts]: bigIntValue*

**Review by mpetrunic — CHANGES_REQUESTED** [review]
- I don't see bn.js dependency removed from lockfile

We should probably bump major version of everything so we know that we transitioned to bigint (cc @GregTheGreek @wemeetagain).

@ec2 did you see any meaningful spec test speed improvement?

File: packages/ssz-cli/.vscode/settings.json (hunk 21 above)
Comment (ec2) [reply packages/ssz-cli/.vscode/settings.json]: WHOOOOPS. accidental commit

File: packages/ssz/src/core/deserialize.ts (hunk 23 above)
Comment (ec2) [reply packages/ssz/src/core/deserialize.ts]: Cool. Ill fix. This was before @wemeetagain blessed me with the bigint-buffer library. Forgot to update here

File: packages/ssz/test/unit/utils.ts (hunk 24 above)
Comment (ec2) [reply packages/ssz/test/unit/utils.ts]: Yes, you need that. JSON doesnt know how to handle bigints.

File: .eslintrc.js (hunk 26 above)
Comment (ec2) [reply .eslintrc.js:65]: I'll add it back. It was giving me weird issues

Comment (ec2) [reply .eslintrc.js:65]: First time messing with the the eslint config 😅

File: packages/lodestar/test/e2e/sync/reqResp.test.ts (hunk 31 above)
Comment (ec2) [reply packages/lodestar/test/e2e/sync/reqResp.test.ts:49]: Not entirely sure... It was a BN, so just changed to bigint haha

File: packages/ssz/test/unit/utils.ts (hunk 24 above)
Comment (mpetrunic) [reply packages/ssz/test/unit/utils.ts]: shouldnt we register it somewhere global?

File: packages/lodestar/src/interop/keypairs.ts (hunk 29 above)
Comment (wemeetagain) [reply packages/lodestar/src/interop/keypairs.ts]: this is the one big endian conversion

Comment (mpetrunic) [reply packages/lodestar/src/interop/keypairs.ts]: oh yeah, thats wierd

File: packages/ssz/test/unit/utils.ts (hunk 24 above)
Comment (wemeetagain) [reply packages/ssz/test/unit/utils.ts]: where are we converting bigints to json?

Comment (mpetrunic) [reply packages/ssz/test/unit/utils.ts]: whenever you are logging some eth2 type?

Comment (wemeetagain) [reply packages/ssz/test/unit/utils.ts]: I don't think we actually do this tho.
I'm wary of attaching custom code to globals. if, for some (sad) reason, a library relies on the default behavior, we're effecting that.

File: packages/lodestar/src/interop/keypairs.ts (hunk 29 above)
Comment (aunyks) [reply packages/lodestar/src/interop/keypairs.ts]: In the future I think it'd be ideal to replace BigInt type operations with constant time operations when we're handling sensitive keys like this.  The `toBufferLE` call that exists in `bytesToBN` [performs a couple String operations in addition to BigInt operations](https://github.com/no2chem/bigint-buffer/blob/76ed004c3a63bb031525f7b9d33ccb1bb24db919/src/index.ts#L25). 

This type of behavior could lead to timing attacks on keys depending on our execution environment (browser or OS), since Strings and [BigInts](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/BigInt) don't operate in constant time and aren't best for cryptography.

Comment (aunyks) [reply packages/lodestar/src/interop/keypairs.ts]: It's an edge case at this point in time, but I think it should be on our radar for the future. If preferred, I can just open a low priority ticket for now.

Comment (mpetrunic) [reply packages/lodestar/src/interop/keypairs.ts]: these are not secure keys, these are generate from 1...100 numbers for interop tests. We are in process of implementing bls specific way of generating secure bls private keys.

Comment (aunyks) [reply packages/lodestar/src/interop/keypairs.ts]: Gotcha. Thanks for the clarification.

**Review by mpetrunic — CHANGES_REQUESTED** [review]
Looks super awesome, could you just remove coverage from lodestar-validator package.

What happens if you remove bn.js from every package or do we wan't bn.js in ssz or somewhere?

**Comment by wemeetagain** [comment]
> What happens if you remove bn.js from every package or do we wan't bn.js in ssz or somewhere?

We could re-add it in a major version bump, but I think its a low priority for now. We would be changing UintType in ssz-type-schema.

**Review by mpetrunic — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]
